### PR TITLE
#10576: generic utility to unsubscribe from broadcast upon drop of the rx-side.

### PR DESCRIPTION
**Pull request by RGafiyatullin**
Addresses #10576.

The main goal of this PR is to introduce a generalized approach to perform unsubscription from broadcasts upon the drop of a receiver.

The two modules were reworked:
- `sc-utils::notification` — a fanout-broadcast used in:
  - `beefy-gadget::notification::{BeefyBestBlockStream, BeefyBestBlockSender}`;
  - `beefy-gadget::notification::{BeefySignedCommitmentStream, BeefySignedCommitmentSender}`;
  - `sc-finality-grandpa::notification::{GrandpaJustificationStream, GrandpaJustificationSender}`;
- `sc-client-api::notifications::{StorageEventStream, StorageNotifications}`.

Both of those modules:
- define a channel with sending- and a receiving-sides;
- have a shared registry to track subscriptions.

The `sc-utils::notification` used to clean up the shared registry upon attempt to send another broadcast-message.
That worked well but it required some liveness on the notification-stream in order to collect the outdated subscriptions.

The `sc-client-api::notifications` was cleaned up both upon failure when trying to send a message, and upon drop of the receiver.
I removed the first cleanup variant, since I could not imagine a situation in which this routine will be effectively executed (therefore could not be cover this code with a test).

> 1) For a producer in order to send a message — the registry must be locked;
>
> 2) The consumer's rx-channel closes on drop.
> 
> 3) The consumer does not expose an API to call `UnboundedReceiver::close(&mut self)`.
>
> 4) Prior to drop the rx-channel, the consumer's `Drop::drop(&mut)` is invoked.
>
> 5) In order to perform the unsubscription, the consumer's Drop implementation locks the registry.
>
> 6) According to **4** and **3**, **2** never happens before to **4** (The locking of the registry in **1** and **5** guarantees the order).

This PR introduces the following:
- module `sc-utils::pubsub`
  - traits `Subscribe<Op>`, `Unsubscribe`, `SubsBase` — to be implemented by a registry;
  - `SharedRegistry<R>` — a wrapper over the registry;
- module `sc-utils::id_sequence` — to generate the sequential subscriber IDs in the registries;

This PR changes the following:
- module `sc-utils::notification`;
- module `sc-client-api::notifications`.

### CODE REVIEW

**Comment by koute** [comment]
Just a note for the future: I see that you took a significant chunk of code essentially unchanged from one file and split it into multiple files (e.g. `tests.rs` or `keys.rs`). It would be significantly easier to review if that was done in a separate commit instead of being also mixed with actual code changes. Right now we basically have a diff here where two files are 100% deleted, and a bunch of files which are 100% new, and when reviewing one has to manually diff the changes between with their eyes them instead of letting the tool do that automatically.

**Comment by RGafiyatullin** [comment]
@koute I've slightly reworked the `sc-utils::pubsub`.
Would you have a look again please?

**Review by bkchr — COMMENTED** [review]
The overall structure of the pr is a little against the rest of the code base.

We don't have these small files that just contain one struct or even files for just implementing traits. I don't want to say that this is not okay some time, but here I don't see any reason for this.

A lot of public methods and types are also missing docs, which makes it a little bit hard to follow what is exactly happening. As @koute also said the pr is a little bit hard to follow because of the splitting of the one file into multiple. 

I did not yet analyzed the entire pr very close, but are there any implementation details besides the removal of this one clean up method?

I also did not really followed the initial discussion about the issue that should be solved, but I would have expected that we also see changes outside of the definition of these utils because we want to simplify other code?

File: client/utils/src/pubsub.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+pub trait Dispatch<K> {
+	type Item;
+	fn dispatch<F>(&mut self, disp_key: K, dispatch: F)
+	where
+		F: FnMut(&SubsID, Self::Item);
+}
+
+#[derive(Debug)]
+pub struct Hub<Ch, R>
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:72]: Added a doc-comment for the structure.

File: client/api/src/notifications.rs [hunk 2]
```diff
@@ -21,30 +21,49 @@
...
 }
 
+/// Manages storage listeners.
+#[derive(Debug)]
+pub struct StorageNotifications<Block: BlockT>(
+	Hub<TracingUnbounded<StorageNotification<Block::Hash>>, Registry>,
+);
+
+/// Type that implements `futures::Stream` of storage change events.
+pub struct StorageEventStream<H>(Receiver<TracingUnbounded<StorageNotification<H>>, Registry>);
```
Comment (bkchr) [inline client/api/src/notifications.rs:62]: Can this not just be a type def?

File: client/api/src/notifications.rs [hunk 3]
```diff
@@ -21,30 +21,49 @@
...
 
+mod registry;
+
+use registry::Registry;
+
+#[cfg(test)]
+mod tests;
+
+/// A type of a message delivered to the subscribers
+pub type StorageNotification<Hash> = (Hash, StorageChangeSet);
```
Comment (bkchr) [inline client/api/src/notifications.rs:44]:
```suggestion
pub struct StorageNotification<Hash> { hash: Hash, changes: StorageChangeSet }
```

Would make it nicer to work with the type.

File: client/api/src/notifications/registry.rs [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+type SubscribersGauge = CounterVec<U64>;
+
+pub(super) struct SubscribeOp<'a> {
+	pub filter_keys: Option<&'a [StorageKey]>,
+	pub filter_child_keys: Option<&'a [(StorageKey, Option<Vec<StorageKey>>)]>,
+}
+
+#[derive(Debug)]
+pub(super) struct Registry {
+	pub(super) metrics: Option<SubscribersGauge>,
```
Comment (bkchr) [inline client/api/src/notifications/registry.rs:37]: Why are all of them `pub(super)`? Do we really need this?

File: client/api/src/notifications/registry.rs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+use sp_core::hexdisplay::HexDisplay;
+
+use fnv::{FnvHashMap, FnvHashSet};
+use prometheus_endpoint::{register, CounterVec, Opts, U64};
+
+use sc_utils::pubsub::{Dispatch, SubsID as SubscriberId, Subscribe, Unsubscribe};
+
+type SubscribersGauge = CounterVec<U64>;
+
+pub(super) struct SubscribeOp<'a> {
```
Comment (bkchr) [inline client/api/src/notifications/registry.rs:30]: Docs please

File: client/api/src/notifications/registry.rs [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+		});
+
+		if let Some(m) = self.metrics.as_ref() {
+			m.with_label_values(&[&"added"]).inc();
+		}
+
+		assert!(self.sinks.insert(subs_id, SubscriberSink::new(subs_id, keys, child_keys)).is_none(), "
+			Each `subs_id` is taken from `self.id_sequence.next_id()`.
+			If we have a duplicate key here, it's either the implementation of `IDSequence` was broken, or we've overflowed `u64`.
+			We are not likely to overflow an `u64`.");
```
Comment (bkchr) [inline client/api/src/notifications/registry.rs:167]: This proof is not correct, because you don't know how `subs_id` was generated.

Yes, I know it is done here in the pr, but that could be broken easily in some later refactoring.

File: client/utils/src/pubsub.rs [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+///
+/// Allows to create a pair of tx and rx, and to send a message over the tx.
+pub trait Channel {
+	/// The sending side of the channel.
+	type Tx;
+
+	/// The receiving side of the channel.
+	type Rx;
+
+	/// The type of an item that can be sent through this channel.
```
Comment (bkchr) [inline client/utils/src/pubsub.rs:92]: Why do we need this?

We currently only have one implementation. Where is the need to have this that generic?

File: client/utils/src/notification.rs [hunk 8]
```diff
@@ -44,108 +48,93 @@ pub trait TracingKeyStr {
...
+		// - once while sending: filter on `!tx.unbounded_send().is_err()`.
+		//
+		// Since there is no `close` or `disconnect` operation defined on the
+		// `NotificationReceiver<Payload>`, the only way to close the `rx` is to drop it.
+		// Upon being dropped the `NotificationReceiver<Payload>` unregisters its `rx`
+		// from the registry using its `_subs_guard`.
+		//
+		// So there's no need to clean up the subscribers set upon sending another message.
+
+		let payload = payload()?; // FIXME: Did it have to be lazily instantiated?
```
Comment (bkchr) [inline client/utils/src/notification.rs:113]: If you read the old code, you see that this was done to not construct the payload when there would be no subscriber. I'm not sure about the costs of the payload, but we should keep it the way.

File: client/api/src/notifications/registry.rs [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+			fmt.debug_map()
+				.entries(map.iter().map(|(key, values)| (HexDisplay::from(key), PrintKeys(values))))
+				.finish()
+		} else {
+			write!(fmt, "None")
+		}
+	}
+}
+
+mod helpers {
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/api/src/notifications/registry.rs]: No need for that indeed.

File: client/api/src/notifications.rs (hunk 2 above)
Comment (RGafiyatullin) [reply client/api/src/notifications.rs:62]: In order to define it as just a type-def I would need to make the `Registry` visible from outside.

The `Registry` has no other need to be exposed.

File: client/api/src/notifications/registry.rs (hunk 4 above)
Comment (RGafiyatullin) [reply client/api/src/notifications/registry.rs:37]: The tests rely on the knowledge of the Registry's internals, `pub(super)` is the minimal visibility that would satisfy that need.

File: client/utils/src/notification.rs (hunk 8 above)
Comment (RGafiyatullin) [reply client/utils/src/notification.rs:113]: Bringing lazy payload materialisation back can be done quite simply: https://github.com/RGafiyatullin/substrate/commit/f51fecc1656f5ffe853ca15795095187979e2102 .


But, in my opinion, providing such API can lead to a surprise when the user isn't duly alert.

So I peeked into the current usages.

Some of the notifications are hashes, I believe those are not tremendously expensive to clone in advance (even if no one is listening):

- https://github.com/paritytech/substrate/blob/master/client/beefy/src/worker.rs#L252
- https://github.com/paritytech/substrate/blob/master/client/beefy/src/worker.rs#L377
- https://github.com/paritytech/substrate/blob/master/client/beefy/rpc/src/lib.rs#L260 (in a test)

Some are structures having a `Vec` field, those could benefit more if they were `Arc`-ed:

- https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/environment.rs#L1334 
- https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/rpc/src/lib.rs#L472 (in a test)
- https://github.com/paritytech/substrate/blob/master/client/beefy/src/worker.rs#L370
- https://github.com/paritytech/substrate/blob/master/client/beefy/rpc/src/lib.rs#L375 (in a test)

File: client/utils/src/pubsub.rs (hunk 7 above)
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs]: I understand that currently the decision on using the unbounded channels has been made.
But if any other policy for handling slow consumers is to be introduced, that would likely to be tried out first on some isolated area: in that case we would just need to implement another type with `Channel` trait defined for it.

Another reason — to have the channel-shuffling code separate from the other logic. It just seemed more neat for me this way. 

If you insist, I will flatten that, so that this trait won't be introduced.

Comment (bkchr) [reply client/utils/src/pubsub.rs]: The plan is to get rid of these unbounded channels, but they will completely disappear. Currently they are just hacks to combine async/sync code. 

Please remove these extra traits etc.

File: client/utils/src/notification.rs (hunk 8 above)
Comment (bkchr) [reply client/utils/src/notification.rs]: Okay, then at least remove the Fixme comment.

Comment (RGafiyatullin) [reply client/utils/src/notification.rs]: For now, I will keep the API as it was before I touched it.

Altering the notification message-types to be `Arc`-ed for the potentially heavy objects would be too much of an invasion for now.

File: client/utils/src/pubsub.rs [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+{
+	shared: Weak<Mutex<Shared<R, Tx>>>,
+	subs_id: SubsID,
+}
+
+#[derive(Debug)]
+struct Shared<R, Tx> {
+	id_sequence: crate::id_sequence::IDSequence,
+	registry: R,
+	sinks: HashMap<SubsID, Tx>,
```
Comment (koute) [inline client/utils/src/pubsub.rs:127]: So this essentially needs to track/manage all of the subscribers twice, right? Once here in this `HashMap`, and once again by whoever's using this. Isn't that somewhat of a half measure? (:

If we don't want to go for what was originally proposed (that is - only adding a channel which notifies the sender out-of-band that the receiver was dropped so that it can remove the subscription) and want to create a more encompassing abstraction which would manage the subscriptions then I think it'd make sense to actually have it fully manage those subscriptions.

**Review by koute — CHANGES_REQUESTED** [review]

File: client/utils/src/pubsub.rs [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+	// NB: this field should be defined after `_unsubs_guard`.
+	rx: TracingUnboundedReceiver<M>,
+}
+
+#[derive(Debug)]
+struct UnsubscribeGuard<R, Tx>
+where
+	R: Unsubscribe,
+{
+	shared: Weak<Mutex<Shared<R, Tx>>>,
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs]: What is the gain in resorting to dynamic dispatch? Just to erase the `R` type out of the receiver part?

I am not convinced that would be better:
- would need to introduce some `UnregisterViaImmutableSelfRef` trait;
- define that trait over `Mutex<Shared<R, ...>> where R: Unsubscribe`;

Having both the existing `Unsubscribe` trait and the `UnregisterViaImmutableSelfRef` trait declared wouldn't be that clear to a reader (in my opinion).

Having a trait `Unsubscribe` to be implemented by the one who implements another pub/sub, while having a `UnregisterViaImmutableSelfRef` declared just as a technicality to perform that dyn-call — it would smell from my point of view.

[the remark this answers is not in the document]
Comment (koute) [reply client/utils/src/pubsub.rs]: > What is the gain in resorting to dynamic dispatch? Just to erase the `R` type out of the receiver part?

Yes.

Because the `Receiver` is a public API, meant to be used outside of this module in many places, and generics in Rust are infectious, so instead of simply writing e.g. `fn foobar<M>(rx: Receiver<M>)` you'll have to write `fn foobar<M, R>(rx: Receiver<M, R>) where R: Unsubscribe` (or have to put the concrete type in) when there's literally no practical benefit in having that `R` there.

I don't think making the public API (mean to be used multiple times in multiple places) more ugly and more inconvenient to use is not a good tradeoff in exchange for just not having to define a single extra *internal* trait in one place.

[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs]: The current usage (and such before this PR) of the notifications introduced a type def for its particular case.

Thus I'd insist, the example of `fn foobar<M, R>(rx: Receiver<M, R>)` is a bit far-fetched. 
Realistically this function is more likely defined as `fn foobar(rx: GrandpaJustificationStream)`.

So the details of "having some registries inside" never leave the region of code where it is actually decided, that some registries are used (e.g. `sc-finality-grandpa::notifications`).


Generics in this particular case do not have to be infectious: one could just define the function as `fn foobar<S>(events_stream) where S: Stream<Item = ANotification>`; in your example — you chose to bring the whole internal machinery into the function's signature, although you did not have to.

File: client/utils/src/pubsub.rs (hunk 10 above)
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs]: Handling the channels externally (in relation to the Registry implementation) indeed introduces another indirection when choosing a channel.

In my opinion that is a reasonable price to keep the Registry implementation unaware of the nature of the channels (or for that matter parametrising the Registry with a type of a channel).

File: client/utils/src/pubsub.rs [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+	id_sequence: crate::id_sequence::IDSequence,
+	registry: R,
+	sinks: HashMap<SeqID, TracingUnboundedSender<M>>,
+}
+
+impl<M, R> AsMut<R> for Shared<M, R> {
+	fn as_mut(&mut self) -> &mut R {
+		&mut self.registry
+	}
+}
```
Comment (bkchr) [inline client/utils/src/pubsub.rs:121]: This is weird to have here. Especially as you just have this for the `lock_registry` method below.

Below you can just change the return type to `impl DerefMut<Target = &`a mut R> + 'a`

And use https://docs.rs/lock_api/0.4.6/lock_api/struct.MutexGuard.html#method.map to map the value.

**Review by bkchr — APPROVED** [review]
Looks good so far. Some last nitpicks, after that it should okay.

File: client/utils/src/pubsub.rs [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+	where
+		R: Dispatch<Trigger, Item = M>,
+	{
+		let mut shared = self.shared.lock();
+		let (registry, sinks) = shared.get_mut();
+
+		registry.dispatch(trigger, |subs_id, item| {
+			if let Some(tx) = sinks.get_mut(&subs_id) {
+				if let Err(send_err) = tx.unbounded_send(item) {
+					log::warn!("Sink with SubsID = {} failed to perform unbounded_send: {} ({} as Dispatch<{}, Item = {}>::dispatch(...))", subs_id, send_err, std::any::type_name::<R>(),
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:191]: The dispatch is executed with the lock over the registry held.
Meanwhile there is a defined order between underlying-channel-drop and un-subscription (from the receiver's perspective).

Hence the warning: if the execution flow got there — something unforeseen is afoot.

File: client/utils/src/pubsub.rs [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+	}
+}
+
+impl<M, R> Unpin for Receiver<M, R> where R: Unsubscribe {}
+
+impl<M, R> Stream for Receiver<M, R>
+where
+	R: Unsubscribe,
+{
+	type Item = <TracingUnboundedReceiver<M> as Stream>::Item;
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:234]: Such way of specifying the type is everywhere in this repository (e.g. [here](https://github.com/paritytech/substrate/blob/dc4cf6c2c73ef5af9197b3e73fb590b1a045d93e/client/network/src/protocol.rs#L1377) ).

My logic was: since the `Stream` implementation for the `Receiver<M,R>` just forwards the logic to the underlying channel, it would only be consistent to forward the type definitions as well.

[the remark this answers is not in the document]
Comment (bkchr) [reply client/utils/src/pubsub.rs:232]: But we know that `M` is the message here. It makes no sense when you directly know the type, to do some extra gymnastics.

File: client/utils/src/pubsub.rs [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+
+/// A subscription hub.
+///
+/// Does the subscription and dispatch.
+/// The exact subscription and routing behaviour is to be implemented by the Registry (of type `R`).
+/// The Hub under the hood uses the channel defined in `crate::mpsc` module.
+#[derive(Debug)]
+pub struct Hub<M, R> {
+	tracing_key: &'static str,
+	shared: Arc<Mutex<Shared<M, R>>>,
```
Comment (koute) [inline client/utils/src/pubsub.rs:92]: If either `Hub::subscribe` or `Hub::send` gets called or another `Receiver` from the same `Hub` is dropped on the same thread from within `Subscribe::subscribe`, `Unsubscribe::unsubscribe` or `Dispatch::dispatch` then this mutex will deadlock.

1. We should detect when that happens and `panic!` instead of deadlocking. (e.g. by keeping a `ThreadLocal` from the [`thread_local`](https://crates.io/crates/thread_local) crate tracking whether the `Mutex` was already locked on the current thread and panicking if it was)
2. It should be documented that calling those methods in those situations is disallowed.
3. A test for the panicking behavior would be nice.

File: client/utils/src/pubsub.rs [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+	pub fn subscribe<K>(&self, subs_key: K) -> Receiver<M, R>
+	where
+		R: Subscribe<K> + Unsubscribe,
+	{
+		let mut shared = self.shared.lock();
+
+		let subs_id = shared.id_sequence.next_id();
+		let (tx, rx) = crate::mpsc::tracing_unbounded(self.tracing_key);
+		assert!(shared.sinks.insert(subs_id, tx).is_none(), "Used IDSequence to create another ID. Should be unique until u64 is overflowed. Should be unique.");
+		shared.registry.subscribe(subs_key, subs_id);
```
Comment (koute) [inline client/utils/src/pubsub.rs:169]: If `subscribe` panics then the `tx` will be leaked in `sinks`.

File: client/utils/src/pubsub.rs [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+	registry: R,
+	sinks: HashMap<SeqID, TracingUnboundedSender<M>>,
+}
+
+impl<M, R> Hub<M, R>
+where
+	R: Unsubscribe,
+{
+	/// Provide mutable access to the registry (for test purposes).
+	pub fn lock_registry<'a>(&'a self) -> impl DerefMut<Target = R> + 'a {
```
Comment (koute) [inline client/utils/src/pubsub.rs:122]: If this is only for testing and is not supposed to be used in normal code then *personally* I'd explicitly name it as such e.g. by putting in the name that it's only for testing, something like `lock_registry_FOR_TESTS`. (Unfortunately we don't do that very often so I'm not going to force it; feel free to ignore this if you want.)

**Review by koute — CHANGES_REQUESTED** [review]

File: client/utils/src/pubsub.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (koute) [reply client/utils/src/pubsub.rs]: I disagree. (: 

But very well, I'm not going to force it.

File: client/utils/src/notification/registry.rs [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+{
+	type Item = Payload;
+	type Ret = Result<(), Error>;
+
+	fn dispatch<F>(&mut self, make_payload: MakePayload, mut dispatch: F) -> Self::Ret
+	where
+		F: FnMut(&SeqID, Self::Item),
+	{
+		if !self.subscribers.is_empty() {
+			let payload = make_payload()?;
```
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/notification/registry.rs:56]: I don't think widening the scope of this PR will help accepting it.
I'd prefer this PR to preserve the existing APIs.

Having the "heavy" notifications Arc-based — is in my opinion something useful still. Can we make that another issue please?

File: client/utils/src/pubsub.rs (hunk 15 above)
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:92]: As I see it, such situation may occur if a Mutex's locked state spans across several `Future::poll` calls.

Every call that touches the Hub's mutex — is a sync-call, i.e. such call shall not return without releasing the mutex.

Did I understand correctly what your concerns were?

Comment (koute) [reply client/utils/src/pubsub.rs:92]: I meant in general, not only when a `Mutex` is locked across multiple `await`s.

If you lock a Mutex then you can't lock it again from the same thread, or it will permanently deadlock. (And deadlocks are not fun to debug.) Currently it's possible to trigger such a deadlock, which isn't great. Instead of allowing for a deadlock we should just detect such cases and panic.

Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:92]: Isn't that a bit overly protective?

That particular mutex is not a part of a public interface<sup>*</sup>, hence neither the outside-user of a `Hub`, nor the implementor of this module's traits can violate the assumptions of how this mutex is used.

If that protective attitude to the private implementation details is a common practice here — I am prepared to do as you tell, of course. 

<sup>*</sup> given that I will mark `fn lock_registry(...)` as `#[cfg(test)]`.

File: client/utils/src/pubsub.rs (hunk 17 above)
Comment (bkchr) [reply client/utils/src/pubsub.rs:122]: Or when we only use this internally for tests (not verified), we can guard it as well.

**Comment by bkchr** [comment]
There are still a lot of comments from me which are not addressed.

File: client/utils/src/pubsub.rs (hunk 15 above)
Comment (koute) [reply client/utils/src/pubsub.rs:91]: > hence neither the outside-user of a `Hub`, nor the implementor of this module's traits can violate the assumptions of how this mutex is used.

That's incorrect. (: `Mutex` assumes that `lock()` will be called only once from a single thread, and that assumption is indirectly also part of our interface here. It can be deadlocked from outside just by calling the public interface in a certain way (even if you'd delete the currently public `lock_registry` that will be possible), so it's not just a private implementation detail, it's a potential footgun.

File: client/utils/src/pubsub.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs]: Thanks for your input though. :)

File: client/utils/src/pubsub.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:193]: @koute in the light of your recent notes on the order between `sinks.remove->registry.unsubscribe` and `registry.subscribe->sinks.insert`, it's reasonable to keep these warnings: the execution would hit here if the registry's `Subscribe<K>` or `Unsubscribe` had panicked.

[the remark this answers is not in the document]
Comment (koute) [reply client/utils/src/pubsub.rs:193]: Applying my recent notes should make this case impossible. (:

File: client/utils/src/pubsub.rs (hunk 15 above)
Comment (bkchr) [reply client/utils/src/pubsub.rs:91]: We can also just use: https://docs.rs/parking_lot/latest/parking_lot/type.ReentrantMutex.html

Comment (koute) [reply client/utils/src/pubsub.rs:91]: > We can also just use: https://docs.rs/parking_lot/latest/parking_lot/type.ReentrantMutex.html

Ah, indeed! That would probably be simpler! :+1:

File: client/utils/src/pubsub.rs (hunk 17 above)
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs]: I decided to opt for 
```rust
pub fn map_registry_for_tests<MapF, Ret>(&self, map: MapF) -> Ret
	where
		MapF: FnOnce(&R) -> Ret
```
as it is a safer thing to have in a public API.

File: client/utils/src/pubsub.rs (hunk 15 above)
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:91]: @koute , @bkchr , would you be able to draw an unambiguous "line" dividing the cases when it is safe to use just a `Mutex` in this project, and when the gymnastics with a `ReentrantMutex` + `RefCell` would be due?

I am not being stubborn, I genuinely do not see a certain way to just call the public API, so that the `Mutex` remains sideways afterwards.

Comment (koute) [reply client/utils/src/pubsub.rs:91]: > would you be able to draw an unambiguous "line" dividing the cases when it is safe to use just a `Mutex` in this project, and when the gymnastics with a `ReentrantMutex` + `RefCell` would be due?

In no case it should be possible to inadvertently deadlock a `Mutex` when you're not explicitly handling a `Mutex`.

In this case, again, you can trigger a deadlock of this internal `Mutex` *from the outside*. You don't have to directly access it. You don't have to even know that it's in there.

> I genuinely do not see a certain way to just call the public API, so that the `Mutex` remains sideways afterwards.

Let me quote myself:

> If either `Hub::subscribe` or `Hub::send` gets called or another `Receiver` from the same `Hub` is dropped on the same thread from within `Subscribe::subscribe`, `Unsubscribe::unsubscribe` or `Dispatch::dispatch` then this mutex will deadlock.

So, for example:

1. A `Receiver` is dropped.
2. `Drop::drop` is called for the `Receiver`.
3. The `Mutex` is locked from inside of `Receiver`'s `drop`.
4. `Shared::unsubscribe` is called while the `Mutex` is being locked.
3. `<R as Unsubscribe>::unsubscribe` is called.
4. The `unsubscribe` impl drops another `Receiver`, either explicitly or entirely by accident due to some weird machination.
5. The `Mutex` is locked again; the previous lock is still in effect.
6. We deadlock.

Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:91]: I am having hard time imagining how step-6 could possibly happen.
It would be required for a Registry to possess an instance of a `Receiver` (of the same hub).
Wouldn't that be exotic to such extent that it would not pass a review such as this one?

Comment (koute) [reply client/utils/src/pubsub.rs:91]: It doesn't have to be directly, it might be indirectly due to some weird indirect machination, e.g. the registry could posses an `Arc` to something, and that object could contain another object, and that another object could contain a `Receiver` (e.g. maybe it would be a list of other receivers or something), and suddenly that `Arc` is the last one out there so the whole thing gets dropped.

It doesn't really matter; what's important is that it's possible to trigger this by mistake. You could use the same argument as for why we might as well use C++ and not Rust: "but writing out of bounds of this array would not pass a review such as this one, so why have bounds checking?". It doesn't matter; the code has to be correct in every circumstance, because sooner or later somebody will screw up. (Same reason why scoped threads were removed from Rust's standard library. You had to really work for it to trigger the unsafety so it was unlikely for anyone to trigger it, but it was unsound so it was removed.)

Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:91]: Okay. I will add the reentrant+refcell thing here,
although I am still convinced that in order to become a footgun that would require quite a lot of dubious practices to be used.

Thanks.

File: client/utils/src/pubsub.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (RGafiyatullin) [reply client/utils/src/pubsub.rs:201]: @koute I've implemented the `ReentrantMutex+RefCell` registry sharing and an exotic receiver-possession scheme in `pubsub::tests::panicking_registry::t02`, in this case that piece of code is executed. 
So I'd keep it a warning rather than an assertion.

File: client/utils/src/pubsub.rs [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+		});
+
+		dispatch_result
+	}
+}
+
+impl<M, R> Shared<M, R> {
+	fn get_mut(&mut self) -> (&mut R, &mut HashMap<SeqID, TracingUnboundedSender<M>>) {
+		(&mut self.registry, &mut self.sinks)
+	}
```
Comment (koute) [inline client/utils/src/pubsub.rs:223]: Any particular reason why you had to add this method? Seems a little unnecessary when you can just access the fields directly (and the method's private anyway, so it's not like this is used to expose it externally). Am I missing something here?

File: client/utils/src/pubsub/tests/normal_operation.rs [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+		assert!(!rx_01.is_terminated());
+		assert_eq!(None, rx_01.next().await);
+		assert!(rx_01.is_terminated());
+	});
+}
+
+#[test]
+fn positive_subs_count_is_correct_upon_drop_of_rxs() {
+	block_on(async {
```
Comment (koute) [inline client/utils/src/pubsub/tests/normal_operation.rs:48]: This shouldn't need `block_on` since there's no async/await inside.

File: client/utils/src/pubsub/tests/normal_operation.rs [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+
+#[test]
+fn positive_rx_receives_relevant_messages_and_terminates_upon_hub_drop() {
+	block_on(async {
```
Comment (koute) [inline client/utils/src/pubsub/tests/normal_operation.rs:23]: Instead of wrapping the whole body with `block_on` use [`#[tokio::test]`](https://docs.rs/tokio/latest/tokio/attr.test.html) instead.

**Review by koute — APPROVED** [review]
Should be fine once the remaining comments are addressed.
